# Homework 7 Emma Alber

#### Conceptual Exercise 3
##### (a) Explain how k-fold cross-validation is implemented.
A dataset is randomly split into k different subsets (aka folds) of equal or approximately equal size. One fold is then used as the testing set (validation set) while the method is trained (fit) on the remaining folds. Next, the MSE of the attained model is computed for the held-out fold. This is then repeated with every fold being the validation set once, which results in k cycles of training and testing. In the end, an average of MSEs over all cycles is computed.
The number k is chosen by the statistician. If k = n, they're basically performing LOOCV, but usually k is smaller than n.  

##### (b) What are the advantages and disadvantages of k-fold crossvalidation relative to:
i. The validation set approach?

Apart from the computational costs being lower, I don't see any advantage of k-fold crossvalidation relative to the validation set approach after reading the book. A disadvantage of the validation set approach compared to k-fold crossvalidation is that yields worse (more variable) estimates for the test error rate, because the validation set approach only makes use of half the dataset available and tends to overestimate the test error rate.

ii. LOOCV?

If k < n, then an advantage of k-fold crossvalidation is that the method poses less computational problems and takes less time, especially if n is very large. Additionally, the variance in k-fold crossvalidation is lower compared to LOOCV. This is conceptually connected to a disadvantage of k-fold crossvalidation: Compared to LOOCV, the bias is higher compared to LOOCV. We see that this is a bias-variance trade off and k must be chosen in order to find the best compromise here.


#### Applied Exercise 7
In Sections 5.3.2 and 5.3.3, we saw that the cv.glm() function can be used in order to compute the LOOCV test error estimate. Alternatively, one could compute those quantities using just the glm() and predict.glm() functions, and a for loop. You will now take this approach in order to compute the LOOCV error for a simple logistic regression model on the Weekly data set. Recall that in the context of classification problems, the LOOCV error is given in (5.4).

a) Fit a logistic regressionmodel that predicts Direction using Lag1 and Lag2.

In [8]:
library("psych")
library("ISLR")
summary(Weekly)

set.seed(1)
attach(Weekly)
fit.glm <- glm(Direction ~ Lag1 + Lag2, data = Weekly, family = "binomial")
summary(fit.glm)

      Year           Lag1               Lag2               Lag3         
 Min.   :1990   Min.   :-18.1950   Min.   :-18.1950   Min.   :-18.1950  
 1st Qu.:1995   1st Qu.: -1.1540   1st Qu.: -1.1540   1st Qu.: -1.1580  
 Median :2000   Median :  0.2410   Median :  0.2410   Median :  0.2410  
 Mean   :2000   Mean   :  0.1506   Mean   :  0.1511   Mean   :  0.1472  
 3rd Qu.:2005   3rd Qu.:  1.4050   3rd Qu.:  1.4090   3rd Qu.:  1.4090  
 Max.   :2010   Max.   : 12.0260   Max.   : 12.0260   Max.   : 12.0260  
      Lag4               Lag5              Volume            Today         
 Min.   :-18.1950   Min.   :-18.1950   Min.   :0.08747   Min.   :-18.1950  
 1st Qu.: -1.1580   1st Qu.: -1.1660   1st Qu.:0.33202   1st Qu.: -1.1540  
 Median :  0.2380   Median :  0.2340   Median :1.00268   Median :  0.2410  
 Mean   :  0.1458   Mean   :  0.1399   Mean   :1.57462   Mean   :  0.1499  
 3rd Qu.:  1.4090   3rd Qu.:  1.4050   3rd Qu.:2.05373   3rd Qu.:  1.4050  
 Max.   : 12.0260   Max.   : 12.0

Year Lag1   Lag2   Lag3   Lag4   Lag5   Volume    Today  Direction
1    1990  0.816  1.572 -3.936 -0.229 -3.484 0.1549760 -0.270 Down     
2    1990 -0.270  0.816  1.572 -3.936 -0.229 0.1485740 -2.576 Down     
3    1990 -2.576 -0.270  0.816  1.572 -3.936 0.1598375  3.514 Up       
4    1990  3.514 -2.576 -0.270  0.816  1.572 0.1616300  0.712 Up       
5    1990  0.712  3.514 -2.576 -0.270  0.816 0.1537280  1.178 Up       
6    1990  1.178  0.712  3.514 -2.576 -0.270 0.1544440 -1.372 Down     
7    1990 -1.372  1.178  0.712  3.514 -2.576 0.1517220  0.807 Up       
8    1990  0.807 -1.372  1.178  0.712  3.514 0.1323100  0.041 Up       
9    1990  0.041  0.807 -1.372  1.178  0.712 0.1439720  1.253 Up       
10   1990  1.253  0.041  0.807 -1.372  1.178 0.1336350 -2.678 Down     
11   1990 -2.678  1.253  0.041  0.807 -1.372 0.1490240 -1.793 Down     
12   1990 -1.793 -2.678  1.253  0.041  0.807 0.1357900  2.820 Up       
13   1990  2.820 -1.793 -2.678  1.253  0.041 0.1398980  4.022 Up       
14   1990  4.022  2.820 -1.793 -2.678  1.253 0.1643420  0.750 Up       
15   1990  0.750  4.022  2.820 -1.793 -2.678 0.1756480 -0.017 Down     
16   1990 -0.017  0.750  4.022  2.820 -1.793 0.1634700  2.420 Up       
17   1990  2.420 -0.017  0.750  4.022  2.820 0.1726250 -1.225 Down     
18   1990 -1.225  2.420 -0.017  0.750  4.022 0.1684460  1.171 Up       
19   1990  1.171 -1.225  2.420 -0.017  0.750 0.1552920 -2.061 Down     
20   1990 -2.061  1.171 -1.225  2.420 -0.017 0.1433920  0.729 Up       
21   1990  0.729 -2.061  1.171 -1.225  2.420 0.1405540  0.112 Up       
22   1990  0.112  0.729 -2.061  1.171 -1.225 0.1250750  2.480 Up       
23   1990  2.480  0.112  0.729 -2.061  1.171 0.1716040 -1.552 Down     
24   1990 -1.552  2.480  0.112  0.729 -2.061 0.1669560 -2.259 Down     
25   1990 -2.259 -1.552  2.480  0.112  0.729 0.1717180 -2.428 Down     
26   1990 -2.428 -2.259 -1.552  2.480  0.112 0.2098160 -2.708 Down     
27   1990 -2.708 -2.428 -2.259 -1.552  2.480 0.1927060 -2.292 Down     
28   1990 -2.292 -2.708 -2.428 -2.259 -1.552 0.1482520 -4.978 Down     
29   1990 -4.978 -2.292 -2.708 -2.428 -2.259 0.1898580  3.547 Up       
30   1990  3.547 -4.978 -2.292 -2.708 -2.428 0.1278840  0.260 Up       
⋮    ⋮    ⋮      ⋮      ⋮      ⋮      ⋮      ⋮         ⋮      ⋮        
1060 2010 -2.252  0.158 -4.226  2.232 -6.388 5.368597   2.509 Up       
1061 2010  2.509 -2.252  0.158 -4.226  2.232 5.369514   2.374 Up       
1062 2010  2.374  2.509 -2.252  0.158 -4.226 4.637208  -3.646 Down     
1063 2010 -3.646  2.374  2.509 -2.252  0.158 4.699712  -5.032 Down     
1064 2010 -5.032 -3.646  2.374  2.509 -2.252 5.100892   5.416 Up       
1065 2010  5.416 -5.032 -3.646  2.374  2.509 4.419372  -1.213 Down     
1066 2010 -1.213  5.416 -5.032 -3.646  2.374 4.487664   3.548 Up       
1067 2010  3.548 -1.213  5.416 -5.032 -3.646 4.580286  -0.096 Down     
1068 2010 -0.096  3.548 -1.213  5.416 -5.032 4.271320   1.819 Up       
1069 2010  1.819 -0.096  3.548 -1.213  5.416 3.963460  -3.779 Down     
1070 2010 -3.779  1.819 -0.096  3.548 -1.213 3.906558  -0.700 Down     
1071 2010 -0.700 -3.779  1.819 -0.096  3.548 3.777406  -0.663 Down     
1072 2010 -0.663 -0.700 -3.779  1.819 -0.096 3.951328   3.750 Up       
1073 2010  3.750 -0.663 -0.700 -3.779  1.819 3.718470   0.456 Up       
1074 2010  0.456  3.750 -0.663 -0.700 -3.779 3.195238   1.446 Up       
1075 2010  1.446  0.456  3.750 -0.663 -0.700 3.972432   2.050 Up       
1076 2010  2.050  1.446  0.456  3.750 -0.663 3.884522  -0.212 Down     
1077 2010 -0.212  2.050  1.446  0.456  3.750 4.037410   1.650 Up       
1078 2010  1.650 -0.212  2.050  1.446  0.456 3.905616   0.948 Up       
1079 2010  0.948  1.650 -0.212  2.050  1.446 4.449160   0.586 Up       
1080 2010  0.586  0.948  1.650 -0.212  2.050 4.576282   0.015 Up       
1081 2010  0.015  0.586  0.948  1.650 -0.212 4.116414   3.599 Up       
1082 2010  3.599  0.015  0.586  0.948  1.650 4.798758  -2.173 Down     
1083 2010 -2.173  3.599  0.015  0.586  0.948 

The following objects are masked from Weekly (pos = 3):

    Direction, Lag1, Lag2, Lag3, Lag4, Lag5, Today, Volume, Year





Call:
glm(formula = Direction ~ Lag1 + Lag2, family = "binomial", data = Weekly)

Deviance Residuals: 
   Min      1Q  Median      3Q     Max  
-1.623  -1.261   1.001   1.083   1.506  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  0.22122    0.06147   3.599 0.000319 ***
Lag1        -0.03872    0.02622  -1.477 0.139672    
Lag2         0.06025    0.02655   2.270 0.023232 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1496.2  on 1088  degrees of freedom
Residual deviance: 1488.2  on 1086  degrees of freedom
AIC: 1494.2

Number of Fisher Scoring iterations: 4


(b) Fit a logistic regressionmodel that predicts Direction using Lag1 and Lag2 using all but the first observation.

In [4]:
fit.glm.1 <- glm(Direction ~ Lag1 + Lag2, data = Weekly[-1, ], family = "binomial")
summary(fit.glm.1)


Call:
glm(formula = Direction ~ Lag1 + Lag2, family = "binomial", data = Weekly[-1, 
    ])

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-1.6258  -1.2617   0.9999   1.0819   1.5071  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  0.22324    0.06150   3.630 0.000283 ***
Lag1        -0.03843    0.02622  -1.466 0.142683    
Lag2         0.06085    0.02656   2.291 0.021971 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1494.6  on 1087  degrees of freedom
Residual deviance: 1486.5  on 1085  degrees of freedom
AIC: 1492.5

Number of Fisher Scoring iterations: 4


(c) Use the model from (b) to predict the direction of the first observation. You can do this by predicting that the first observation will go up if P(Direction="Up"|Lag1, Lag2) > 0.5. Was this observation correctly classified?

In [9]:
predict.glm(fit.glm.1, Weekly[1, ], type = "response") > 0.5
head(Weekly)

1 
TRUE

Year Lag1   Lag2   Lag3   Lag4   Lag5   Volume    Today  Direction
1 1990  0.816  1.572 -3.936 -0.229 -3.484 0.1549760 -0.270 Down     
2 1990 -0.270  0.816  1.572 -3.936 -0.229 0.1485740 -2.576 Down     
3 1990 -2.576 -0.270  0.816  1.572 -3.936 0.1598375  3.514 Up       
4 1990  3.514 -2.576 -0.270  0.816  1.572 0.1616300  0.712 Up       
5 1990  0.712  3.514 -2.576 -0.270  0.816 0.1537280  1.178 Up       
6 1990  1.178  0.712  3.514 -2.576 -0.270 0.1544440 -1.372 Down

The true direction is "down", but the output shows us that it was classified as "up". The observation was not classified correctly. 

(d) Write a for loop from i = 1 to i = n, where n is the number of observations in the data set, that performs each of the following steps:

i. Fit a logistic regression model using all but the ith observation to predict Direction using Lag1 and Lag2.

ii. Compute the posterior probability of the market moving up for the ith observation.

iii. Use the posterior probability for the ith observation in order to predict whether or not the market moves up.

iv. Determine whether or not an error was made in predicting the direction for the ith observation. If an error was made, then indicate this as a 1, and otherwise indicate it as a 0.

In [11]:
error <- rep(0, dim(Weekly)[1])
for (i in 1:dim(Weekly)[1]) 
{fit.glm <- glm(Direction ~ Lag1 + Lag2, data = Weekly[-i, ],  family = "binomial")
    pred.up <- predict.glm(fit.glm, Weekly[i, ], type = "response") > 0.5
    true.up <- Weekly[i, ]$Direction == "Up"
    if (pred.up != true.up)
        error[i] <- 1}

error
sum(error)

[1] 1 1 0 1 0 1 0 0 0 1 1 0 1 0 1 0 1 0 1 0 0 0 1 1 1 1 1 1 0 1 1 1 1 0 1 0 0
  [38] 0 1 0 1 0 0 1 0 1 1 1 0 1 0 0 0 1 0 0 1 1 0 0 0 0 1 0 1 1 0 0 1 0 1 1 0 0
  [75] 0 1 0 1 1 0 0 1 1 0 1 1 0 0 1 0 0 1 1 1 0 0 0 0 0 1 0 1 1 0 0 1 0 1 0 0 1
 [112] 1 0 0 1 0 0 1 0 0 1 1 1 1 0 0 0 1 0 1 0 1 1 0 0 0 1 1 1 0 0 0 1 0 0 0 0 0
 [149] 0 1 1 1 0 1 0 0 1 1 0 1 0 0 1 1 0 0 1 0 0 1 0 0 1 1 1 0 1 0 1 0 0 0 0 0 0
 [186] 0 0 1 1 0 1 0 1 0 1 0 1 0 0 1 0 0 1 0 0 1 0 1 0 1 1 1 0 0 1 1 0 1 0 0 1 1
 [223] 0 0 0 1 1 1 0 1 0 1 0 1 0 0 0 1 1 0 1 0 1 0 1 0 1 0 1 1 0 1 0 0 1 0 0 1 0
 [260] 0 0 0 0 1 0 0 0 1 0 0 1 0 0 0 1 0 0 1 0 0 1 0 0 1 0 1 1 0 0 0 0 0 1 0 1 0
 [297] 0 1 0 0 0 1 0 0 1 1 0 0 1 0 0 0 0 1 0 1 1 0 0 1 0 1 0 1 1 0 0 0 1 0 1 0 0
 [334] 1 1 1 1 0 1 0 0 1 0 0 0 1 0 1 0 1 0 0 0 0 0 1 1 0 0 1 0 0 1 0 0 0 1 1 0 1
 [371] 1 1 1 1 0 0 0 1 0 0 0 0 0 0 1 0 1 1 0 0 1 1 0 0 0 0 0 1 0 0 1 1 1 0 1 0 1
 [408] 0 1 1 1 0 1 0 0 0 0 0 0 0 0 0 0 1 0 1 0 1 0 1 0 0 1 0 1 0 0 0 0 0 1 1 1 1
 [445] 0 1 1 0 1 0 1 1 0 1 0 0 1 0 0 1 1 0 0 0 0 1 1 0 0 1 0 1 0 0 0 1 0 0 1 0 0
 [482] 0 1 1 1 0 1 0 0 0 1 0 1 1 1 0 0 0 0 1 1 1 0 1 1 0 1 0 0 0 1 0 1 0 0 0 1 0
 [519] 1 1 0 0 1 1 0 0 0 1 1 0 1 0 1 1 1 1 1 0 0 0 1 0 0 0 1 1 0 1 0 0 0 1 1 1 1
 [556] 1 1 0 1 0 1 0 0 1 0 0 1 1 1 0 0 0 1 1 1 1 1 1 1 1 1 1 0 1 0 0 1 0 0 1 0 1
 [593] 0 0 1 0 0 1 0 1 1 0 1 1 1 0 1 0 1 0 1 0 0 0 1 0 1 0 1 0 1 1 0 1 1 0 1 0 1
 [630] 0 1 1 1 1 0 1 1 0 0 0 1 1 1 1 0 1 1 1 0 1 0 0 0 1 1 1 1 1 1 0 1 0 0 1 0 0
 [667] 0 1 1 0 1 0 1 1 1 1 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 0 1 0 1 1
 [704] 0 0 0 0 1 0 1 0 1 0 1 0 0 1 1 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 1 1 1 0 1 1 0
 [741] 1 1 1 1 0 0 0 1 1 1 1 1 1 0 1 0 0 0 0 0 0 1 0 1 1 1 0 0 0 1 0 0 1 0 0 0 1
 [778] 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 1 0 0 1 0 0 1 0 0 0 0 0 1 0 1 1 0 0 1 1
 [815] 0 1 1 1 0 0 0 0 0 1 1 0 0 1 0 0 1 0 1 0 0 0 1 1 0 1 1 0 1 0 1 0 1 1 0 0 1
 [852] 1 1 0 1 1 0 0 0 1 0 1 0 1 0 1 0 0 0 0 0 1 0 0 1 1 0 0 1 0 1 0 1 1 0 1 0 1
 [889] 1 0 0 0 1 0 0 0 0 0 0 0 1 0 1 0 1 0 0 0 1 1 1 1 1 0 1 1 0 0 0 0 0 1 0 0 1
 [926] 0 0 0 0 1 0 1 1 1 0 0 1 1 0 1 1 1 1 0 1 0 1 0 1 0 1 0 1 0 0 1 1 1 1 1 0 1
 [963] 0 0 0 1 1 1 0 1 1 1 1 0 0 0 0 1 1 0 0 0 0 1 0 0 1 1 1 0 0 1 1 1 0 1 0 0 0
[1000] 0 1 0 0 1 0 1 0 0 1 1 1 1 0 1 0 0 1 0 0 1 0 0 1 1 0 1 1 1 0 1 1 0 0 0 1 0
[1037] 1 0 1 1 1 1 0 0 1 0 0 0 0 0 0 1 0 1 1 0 0 0 0 0 1 1 1 0 0 0 1 0 1 1 1 0 0
[1074] 0 0 1 0 0 0 0 0 1 0 1 0 0 0 0 0

[1] 490

(e) Take the average of the n numbers obtained in (d) in order to obtain the LOOCV estimate for the test error. Comment on the results.

In [12]:
mean(error)

[1] 0.4499541

By using LOOCV, we get a test error rate of approx. 45%. This means that our model classified about 55% of the test data (which it hadn't been trained on) correctly.

#### Applied Exercise 9
We will now consider the Boston housing data set, from the MASS library.

In [13]:
library(MASS)
Boston
summary(Boston)

crim     zn   indus chas nox   rm    age   dis    rad tax ptratio black 
1   0.00632  18.0 2.31  0    0.538 6.575  65.2 4.0900 1   296 15.3    396.90
2   0.02731   0.0 7.07  0    0.469 6.421  78.9 4.9671 2   242 17.8    396.90
3   0.02729   0.0 7.07  0    0.469 7.185  61.1 4.9671 2   242 17.8    392.83
4   0.03237   0.0 2.18  0    0.458 6.998  45.8 6.0622 3   222 18.7    394.63
5   0.06905   0.0 2.18  0    0.458 7.147  54.2 6.0622 3   222 18.7    396.90
6   0.02985   0.0 2.18  0    0.458 6.430  58.7 6.0622 3   222 18.7    394.12
7   0.08829  12.5 7.87  0    0.524 6.012  66.6 5.5605 5   311 15.2    395.60
8   0.14455  12.5 7.87  0    0.524 6.172  96.1 5.9505 5   311 15.2    396.90
9   0.21124  12.5 7.87  0    0.524 5.631 100.0 6.0821 5   311 15.2    386.63
10  0.17004  12.5 7.87  0    0.524 6.004  85.9 6.5921 5   311 15.2    386.71
11  0.22489  12.5 7.87  0    0.524 6.377  94.3 6.3467 5   311 15.2    392.52
12  0.11747  12.5 7.87  0    0.524 6.009  82.9 6.2267 5   311 15.2    396.90
13  0.09378  12.5 7.87  0    0.524 5.889  39.0 5.4509 5   311 15.2    390.50
14  0.62976   0.0 8.14  0    0.538 5.949  61.8 4.7075 4   307 21.0    396.90
15  0.63796   0.0 8.14  0    0.538 6.096  84.5 4.4619 4   307 21.0    380.02
16  0.62739   0.0 8.14  0    0.538 5.834  56.5 4.4986 4   307 21.0    395.62
17  1.05393   0.0 8.14  0    0.538 5.935  29.3 4.4986 4   307 21.0    386.85
18  0.78420   0.0 8.14  0    0.538 5.990  81.7 4.2579 4   307 21.0    386.75
19  0.80271   0.0 8.14  0    0.538 5.456  36.6 3.7965 4   307 21.0    288.99
20  0.72580   0.0 8.14  0    0.538 5.727  69.5 3.7965 4   307 21.0    390.95
21  1.25179   0.0 8.14  0    0.538 5.570  98.1 3.7979 4   307 21.0    376.57
22  0.85204   0.0 8.14  0    0.538 5.965  89.2 4.0123 4   307 21.0    392.53
23  1.23247   0.0 8.14  0    0.538 6.142  91.7 3.9769 4   307 21.0    396.90
24  0.98843   0.0 8.14  0    0.538 5.813 100.0 4.0952 4   307 21.0    394.54
25  0.75026   0.0 8.14  0    0.538 5.924  94.1 4.3996 4   307 21.0    394.33
26  0.84054   0.0 8.14  0    0.538 5.599  85.7 4.4546 4   307 21.0    303.42
27  0.67191   0.0 8.14  0    0.538 5.813  90.3 4.6820 4   307 21.0    376.88
28  0.95577   0.0 8.14  0    0.538 6.047  88.8 4.4534 4   307 21.0    306.38
29  0.77299   0.0 8.14  0    0.538 6.495  94.4 4.4547 4   307 21.0    387.94
30  1.00245   0.0 8.14  0    0.538 6.674  87.3 4.2390 4   307 21.0    380.23
⋮   ⋮        ⋮    ⋮     ⋮    ⋮     ⋮     ⋮     ⋮      ⋮   ⋮   ⋮       ⋮     
477  4.87141 0    18.10 0    0.614 6.484 93.6  2.3053 24  666 20.2    396.21
478 15.02340 0    18.10 0    0.614 5.304 97.3  2.1007 24  666 20.2    349.48
479 10.23300 0    18.10 0    0.614 6.185 96.7  2.1705 24  666 20.2    379.70
480 14.33370 0    18.10 0    0.614 6.229 88.0  1.9512 24  666 20.2    383.32
481  5.82401 0    18.10 0    0.532 6.242 64.7  3.4242 24  666 20.2    396.90
482  5.70818 0    18.10 0    0.532 6.750 74.9  3.3317 24  666 20.2    393.07
483  5.73116 0    18.10 0    0.532 7.061 77.0  3.4106 24  666 20.2    395.28
484  2.81838 0    18.10 0    0.532 5.762 40.3  4.0983 24  666 20.2    392.92
485  2.37857 0    18.10 0    0.583 5.871 41.9  3.7240 24  666 20.2    370.73
486  3.67367 0    18.10 0    0.583 6.312 51.9  3.9917 24  666 20.2    388.62
487  5.69175 0    18.10 0    0.583 6.114 79.8  3.5459 24  666 20.2    392.68
488  4.83567 0    18.10 0    0.583 5.905 53.2  3.1523 24  666 20.2    388.22
489  0.15086 0    27.74 0    0.609 5.454 92.7  1.8209  4  711 20.1    395.09
490  0.18337 0    27.74 0    0.609 5.414 98.3  1.7554  4  711 20.1    344.05
491  0.20746 0    27.74 0    0.609 5.093 98.0  1.8226  4  711 20.1    318.43
492  0.10574 0    27.74 0    0.609 5.983 98.8  1.8681  4  711 20.1    390.11
493  0.11132 0    27.74 0    0.609 5.983 83.5  2.1099  4  711 20.1    396.90
494  0.17331 0     9.69 0    0.585 5.707 54.0  2.3817  6  391 19.2    396.90
495  0.27957 0     9.69 0    0.585 5.926 42.6  2.3817  6  391 19.2    396.90
496  0.17899 0     9.69 0    0.585 5.670 28.8  2.7986  6  391 19.2    393.29


      crim                zn             indus            chas        
 Min.   : 0.00632   Min.   :  0.00   Min.   : 0.46   Min.   :0.00000  
 1st Qu.: 0.08205   1st Qu.:  0.00   1st Qu.: 5.19   1st Qu.:0.00000  
 Median : 0.25651   Median :  0.00   Median : 9.69   Median :0.00000  
 Mean   : 3.61352   Mean   : 11.36   Mean   :11.14   Mean   :0.06917  
 3rd Qu.: 3.67708   3rd Qu.: 12.50   3rd Qu.:18.10   3rd Qu.:0.00000  
 Max.   :88.97620   Max.   :100.00   Max.   :27.74   Max.   :1.00000  
      nox               rm             age              dis        
 Min.   :0.3850   Min.   :3.561   Min.   :  2.90   Min.   : 1.130  
 1st Qu.:0.4490   1st Qu.:5.886   1st Qu.: 45.02   1st Qu.: 2.100  
 Median :0.5380   Median :6.208   Median : 77.50   Median : 3.207  
 Mean   :0.5547   Mean   :6.285   Mean   : 68.57   Mean   : 3.795  
 3rd Qu.:0.6240   3rd Qu.:6.623   3rd Qu.: 94.08   3rd Qu.: 5.188  
 Max.   :0.8710   Max.   :8.780   Max.   :100.00   Max.   :12.127  
      rad              tax 

(a) Based on this data set, provide an estimate for the population mean of medv. Call this estimate ˆμ.

In [36]:
#i'll call it mü because i don't feel like typing that symbol every time
attach(Boston)
mü_est <- mean(medv, na.rm = TRUE)
mü_est

The following objects are masked from Boston (pos = 3):

    age, black, chas, crim, dis, indus, lstat, medv, nox, ptratio, rad,
    rm, tax, zn


The following objects are masked from Boston (pos = 4):

    age, black, chas, crim, dis, indus, lstat, medv, nox, ptratio, rad,
    rm, tax, zn


The following objects are masked from Boston (pos = 5):

    age, black, chas, crim, dis, indus, lstat, medv, nox, ptratio, rad,
    rm, tax, zn


The following objects are masked from Boston (pos = 6):

    age, black, chas, crim, dis, indus, lstat, medv, nox, ptratio, rad,
    rm, tax, zn


The following objects are masked from Boston (pos = 7):

    age, black, chas, crim, dis, indus, lstat, medv, nox, ptratio, rad,
    rm, tax, zn




[1] 22.53281

(b) Provide an estimate of the standard error of ˆμ. Interpret this result.
Hint: We can compute the standard error of the sample mean by dividing the sample standard deviation by the square root of the number of observations.

In [44]:
a <- sqrt(length(medv))
a
mü_est_sd <- sd(medv)/a
mü_est_sd

[1] 22.49444

[1] 0.4088611

(c) Now estimate the standard error of ˆμ using the bootstrap. How does this compare to your answer from (b)?

In [46]:
library(boot)

boot.fn = function(data, index) return(mean(data[index]))
bootstrap = boot(medv, boot.fn, 1000)
bootstrap


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = medv, statistic = boot.fn, R = 1000)


Bootstrap Statistics :
    original      bias    std. error
t1* 22.53281 -0.02542806   0.4023801

The standard error in the bootstrap-output is very similar to the one attained before (both are approx. 0.41).

(d) Based on your bootstrap estimate from (c), provide a 95% confidence interval for the mean of medv. Compare it to the results obtained using t.test(Boston$medv).
Hint: You can approximate a 95 percent confidence interval using the formula [ˆμ − 2SE(ˆμ), ˆμ + 2SE(ˆμ)].

In [48]:
CI_mü_est <- c(22.53 - 2 * 0.4119, 22.53 + 2 * 0.4119)
CI_mü_est

t.test(Boston$medv)

[1] 21.7062 23.3538


	One Sample t-test

data:  Boston$medv
t = 55.111, df = 505, p-value < 2.2e-16
alternative hypothesis: true mean is not equal to 0
95 percent confidence interval:
 21.72953 23.33608
sample estimates:
mean of x 
 22.53281 


The CI obtained by the t-test is very similar to the on obtained by computing it by hand (both are approx. 21.07 - 23.3)

(e) Based on this data set, provide an estimate, ˆμmed, for the median value of medv in the population.

In [49]:
mü_med_est <- median(medv)
mü_med_est

[1] 21.2

(f) We now would like to estimate the standard error of ˆμmed. Unfortunately, there is no simple formula for computing the standard error of the median. Instead, estimate the standard error of the median using the bootstrap. Comment on your findings.

In [51]:
boot.fn = function(data, index) return(median(data[index]))
boot(medv, boot.fn, 1000)


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = medv, statistic = boot.fn, R = 1000)


Bootstrap Statistics :
    original   bias    std. error
t1*     21.2 -0.00865   0.3809761

The output shows us that we get an estimated median of 21.2, which is the same es I estimated above. The standard error is .38.

(g) Based on this data set, provide an estimate for the tenth percentile of medv in Boston suburbs. Call this quantity ˆμ0.1. (You can use the quantile() function.)

In [52]:
mü_est_0.1 <- quantile(medv, c(0.1))
mü_est_0.1

10% 
12.75

(h) Use the bootstrap to estimate the standard error of ˆμ0.1. Comment on your findings.

In [53]:
boot.fn = function(data, index) return(quantile(data[index], c(0.1)))
boot(medv, boot.fn, 1000)


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = medv, statistic = boot.fn, R = 1000)


Bootstrap Statistics :
    original  bias    std. error
t1*    12.75  0.0176   0.5059657

Just like in f), we see that the estimated value of the 10th percentile is the same as the one obtained by the bootstrap method (12.75). The standard error is 0.51.